### PR TITLE
CO-1548 : exchange of working days in attendances - beta

### PR DESCRIPTION
**Pull request by samuel-fringeli**
view for forced hours and manager messages not created yet

### CODE REVIEW

File: hr_attendance_calendar/models/hr_attendance_day.py [hunk 1]
```diff
@@ -58,6 +58,10 @@ class HrAttendanceDay(models.Model):
     # Attendances
     attendance_ids = fields.One2many('hr.attendance', 'attendance_day_id',
                                      'Attendances', readonly=True)
+    has_logged_hours = fields.Boolean('Has logged hours',
+                                      compute='_compute_has_logged_hours')
+    has_linked_change_day_request = fields.Boolean(
+        default=False, compute='_compute_has_linked_change_day_request')
```
Comment (ecino) [inline hr_attendance_calendar/models/hr_attendance_day.py]: default is useless for a compute field

File: hr_attendance_calendar/models/hr_attendance_day.py [hunk 2]
```diff
@@ -96,6 +100,27 @@ class HrAttendanceDay(models.Model):
     ##########################################################################
     #                             FIELDS METHODS                             #
     ##########################################################################
+    @api.multi
+    @api.depends('attendance_ids')
+    def _compute_has_logged_hours(self):
+        for att_day in self:
+            # > 1 because when we calculate this, we're creating an attendance
```
Comment (ecino) [inline hr_attendance_calendar/models/hr_attendance_day.py]: Are you sure this is still the case? Normally the compute method is called when the field is accessed for the first time (when the field is not stored).

Comment (ecino) [reply hr_attendance_calendar/models/hr_attendance_day.py]: I feel this field is useless, when needed you can simply do `if att_day.attendance_ids` to check if there are logged hours

File: hr_attendance_calendar/models/hr_attendance_day.py [hunk 3]
```diff
@@ -96,6 +100,27 @@ class HrAttendanceDay(models.Model):
...
+
+    @api.multi
+    def get_related_forced_due_hours(self):
+        self.ensure_one()
+        return self.env['hr.forced.due.hours'].search([
+            ('employee_id', '=', self.employee_id.id),
+            ('date', '=', self.date)])
+
+    @api.multi
+    @api.depends('due_hours')
```
Comment (ecino) [inline hr_attendance_calendar/models/hr_attendance_day.py:110]: depends is useless when compute field is not stored

File: hr_attendance_calendar/models/hr_attendance_day.py [hunk 4]
```diff
@@ -126,6 +151,14 @@ def _compute_due_hours(self):
         some hours if they are vacation"""
         for att_day in self:
 
+            # Forced due hours (when an user changes work days)
+            res = att_day.get_related_forced_due_hours()
```
Comment (ecino) [inline hr_attendance_calendar/models/hr_attendance_day.py]: forced_hours would be better name than res

File: hr_attendance_calendar/models/hr_attendance_day.py [hunk 5]
```diff
@@ -126,6 +151,14 @@ def _compute_due_hours(self):
...
+            # res.forced_due_hours returns False if exists and has value 0.0
+            # that's why we use len(res) == 1
+            if len(res) == 1:
```
Comment (ecino) [inline hr_attendance_calendar/models/hr_attendance_day.py]: Your comment is not correct. Here you can simply do `if forced_hours`

**Review by ecino — CHANGES_REQUESTED** [review]
I've not had time to do the full review. Here are some first comments.

File: hr_attendance_calendar/models/hr_forced_due_hours.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    date = fields.Date('Date', required=True)
+    forced_due_hours = fields.Float('Forced due hours', required=True)
+
+    @api.model
+    def recompute_due_hours(self, employee_id, date):
+        return self.env['hr.attendance.day'].search(
+            [('employee_id', '=', employee_id),
+             ('date', '=', date)]).recompute_due_hours()
+
+    @api.one
```
Comment (ecino) [inline hr_attendance_calendar/models/hr_forced_due_hours.py]: api.one is deprecated and should not be used, especially not for write methods

File: hr_attendance_calendar/models/hr_forced_due_hours.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    def write(self, vals):
+        old_vals = dict()
+        for field in ['employee_id', 'date', 'forced_due_hours']:
+            old_vals[field] = getattr(self, field)
+
+        res = super(HrForcedDueHours, self).write(vals)
+
+        self.recompute_due_hours(
+            old_vals['employee_id'].id, old_vals['date'])
```
Comment (ecino) [inline hr_attendance_calendar/models/hr_forced_due_hours.py]: old_vals['employee_id'] should already contain an id and not a record, so remove `.id`

File: hr_attendance_calendar/models/hr_attendance_day.py (hunk 2 above)
Comment (samuel-fringeli) [reply hr_attendance_calendar/models/hr_attendance_day.py]: The field is not used in hr.attendance.day model, but a related field of this is needed in hr.attendance, to show the wizard only if there are already logged hours.
And for len > 1, the field depends on attendance_ids, so it's recomputed each time an attendance_day has a new attendance

File: hr_attendance_calendar/models/hr_forced_due_hours.py (hunk 7 above)
Comment (samuel-fringeli) [reply hr_attendance_calendar/models/hr_forced_due_hours.py]: No, because `old_vals['employee_id'] = self.employee_id` that is an `hr.employee()` object

Comment (ecino) [reply hr_attendance_calendar/models/hr_forced_due_hours.py]: If you do that it's wrong! The write method never takes records as parameters in the dictionary. It should always be ids.

File: hr_attendance_calendar/wizards/hr_change_day_wizard.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        for index, date in enumerate(dates):
+            #  only get days with due hours
+            if due_hours[index] != 0:
+                data[date] = due_hours[index]
+
+        #  delete created att_days
+        unlinked = [x.unlink() for x in att_day_created].count(False) == 0
+
+        if unlinked:
+            return data
```
Comment (ecino) [inline hr_attendance_calendar/wizards/hr_change_day_wizard.py]: Why not returning data in any cases?

**Comment by ecino** [comment]
Please rebase on devel and fix any pylint and flake8 issues

File: hr_attendance_calendar/models/hr_forced_due_hours.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            super(HrForcedDueHours, record).write(vals)
+
+            record.recompute_due_hours(
+                old_vals['employee_id'].id, old_vals['date'])
+            record.recompute_due_hours(record.employee_id.id, record.date)
+
+    @api.multi
+    def unlink(self):
+        length = len(self)
+        employee_ids = [0]*length
```
Comment (ecino) [inline hr_attendance_calendar/models/hr_forced_due_hours.py]: Better, you can do `employee_ids = self.mapped('employee_id.id')` and dates = `self.mapped('date')`

**Review by ecino — CHANGES_REQUESTED** [review]

**Review by eicher31 — APPROVED** [review]

**Review by ecino — APPROVED** [review]

**Pull request merged**